runtime/src/iree/base/internal/math.h

```diff
@@ -299,51 +300,38 @@ static inline float iree_math_make_f32_from_bits(uint32_t src, int exp_bits,
                                                  int bias_tweak,
                                                  bool nan_as_neg_zero) {
   IREE_MATH_FP_FORMAT_CONSTANTS(src_, exp_bits, mantissa_bits, bias_tweak)
-  IREE_MATH_FP_FORMAT_CONSTANTS(f32_, 8, 23, 0)
-  const uint32_t src_sign = src & src_sign_mask;
-  const uint32_t f32_sign = src_sign << (f32_sign_shift - src_sign_shift);
+  const float float_sign = (src & src_sign_mask) ? -1.f : 1.f;
   const uint32_t src_exp = src & src_exp_mask;
   const uint32_t src_mantissa = src & src_mantissa_mask;
-  uint32_t f32_exp = 0;
-  uint32_t f32_mantissa = 0;
   if (src_exp == src_exp_mask) {
     // Top exponent value normally means infinity or NaN.
     if (have_infinity) {
       // NaN or Inf case.
-      f32_exp = f32_exp_mask;
       if (src_mantissa) {
-        f32_mantissa = f32_mantissa_mask;  // Quiet NaN.
+        return NAN;
       } else {
-        f32_mantissa = 0;  // Inf.
+        return float_sign * INFINITY;
       }
     } else {
       // No infinities => more large finite values, unless this is a NaN.
-      bool is_finite = src_mantissa != src_mantissa_mask || nan_as_neg_zero;
-      if (is_finite) {
-        f32_exp = ((src_exp >> src_exp_shift) + f32_exp_bias - src_exp_bias)
-                  << f32_exp_shift;
-        f32_mantissa = src_mantissa << (f32_mantissa_bits - src_mantissa_bits);
-      } else {
-        // NaN. Generate a quiet NaN.
-        f32_exp = f32_exp_mask;
-        f32_mantissa = f32_mantissa_mask;
+      if (src_mantissa == src_mantissa_mask && !nan_as_neg_zero) {
```

Comment:
Hold on - I'm pretty sure any non-zero mantissa is NaN? Or am I misunderstanding the condition

Reply:
Here we are in the `else` branch below the above `if (have_infinity)` so we know that `have_infinity == false`, meaning that the type is a FN type, where the top exponent value is used for additional finite values, outside of the specific case where all mantissa bits are 1 in which case it's a NaN, unless the type is UZ (indicated by `nan_as_neg_zero`) in which case even that value is finite.